subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java

```diff
@@ -106,15 +107,18 @@
     private final Map<Pair<Node, Node>, Boolean> reachableCache = Maps.newHashMap();
     private final Set<Node> dependenciesCompleteCache = Sets.newHashSet();
     private final WorkerLeaseService workerLeaseService;
+    private final SharedResourceLeaseLockRegistry sharedResourceLeaseLockRegistry;
+    private final Map<Node, List<ResourceLock>> sharedResourceLocks = Maps.newIdentityHashMap();
```

Comment:
The issue with caching these locks is that they include a reference to the current worker thread which may change since a different worker might come along and request to execute this node than the original thread that first visited it.

The only way I see around this would be to make the owner thread on `SharedResourceLease` mutable in some way.

Reply:
Actually, there's really no need to pass the worker thread on lock construction. We only care about the owner thread when the lock is acquired so we can just set it then. This simplifies thing a bit and allows us to pre-compute the shared resource locks similarly to how we do for project locks. I've gone ahead and refactored this.